# Universal Dependencies Treebank analysis

Analysis of conllu files, feature analysis and count for NN feature decision

In [1]:
import os
import sys
import numpy as np
import pandas as pd
import conllu



In [2]:
base_dir = "/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4"

In [3]:
for root, fdir, files in os.walk(base_dir):
        print (root)
#         for items in fdir:  #fnmatch.filter(files, "*"):
#                 print (items)


/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Basque-BDT
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Danish-DDT
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Galician-TreeGal
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Turkish-PUD
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Maltese-MUDT
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Irish-IDT
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Galician-CTG
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Komi_Zyrian-Lattice
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Hindi_English-HIENCS
/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Hindi_English-HIENCS/merge
/home/leo/pr

In [4]:
root

'/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Dutch-Alpino'

In [5]:
def get_all_files_recurse(rootdir):
    allfiles = []
    for root, directories, filenames in os.walk(rootdir):
        for filename in filenames: 
            allfiles.append(os.path.join(root,filename) )
    return allfiles

In [6]:
allfiles = get_all_files_recurse(base_dir)

In [7]:
conllufiles = [f for f in allfiles if f.endswith(".conllu")]

In [8]:
len(conllufiles)

339

In [9]:
conllufiles[:5]

['/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Basque-BDT/eu_bdt-ud-dev.conllu',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Basque-BDT/eu_bdt-ud-test.conllu',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Basque-BDT/eu_bdt-ud-train.conllu',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Danish-DDT/da_ddt-ud-test.conllu',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Danish-DDT/da_ddt-ud-train.conllu']

In [10]:
# proces all conllufiles to get a clean file to be able to start doing some stats and get clean representations
cleanfnames = []
cmd = "sed -e 's/#.*$//' -e '/^$/d' {} > {}"
for cf in conllufiles:
    nf = cf.replace(".conllu", ".clean")
    cleanfnames.append(nf)
    tmp = cmd.format(cf, nf)
#     print(tmp)
    os.system(tmp)


In [11]:
len(cleanfnames)

339

In [12]:
# check that all are files
ff = [os.path.isfile(f) for f in cleanfnames]

In [13]:
False in ff

False

In [14]:
colnames = ['ID', 'FORM', 'LEMMA', 'UPOS', 'XPOS', 'FEATS', 'HEAD', 'DEPREL', 'DEPS', 'MISC']

In [15]:

f0 = pd.read_csv(cleanfnames[0], sep='\t', names=colnames)


In [16]:
f0.columns

Index(['ID', 'FORM', 'LEMMA', 'UPOS', 'XPOS', 'FEATS', 'HEAD', 'DEPREL',
       'DEPS', 'MISC'],
      dtype='object')

In [17]:
f0.UPOS.value_counts()

NOUN     5898
VERB     3719
PUNCT    3617
AUX      2512
PROPN    1990
ADJ      1220
CCONJ    1210
ADV       984
NUM       910
DET       820
ADP       365
_         356
PART      292
PRON      164
X          29
INTJ        6
SYM         3
Name: UPOS, dtype: int64

In [18]:
for n in colnames:
    print("------------------")
    print(f0[n].value_counts())

------------------
1     1798
2     1798
3     1798
4     1788
5     1759
6     1677
7     1569
8     1456
9     1327
10    1202
11    1085
12     973
13     856
14     755
15     674
16     582
17     490
18     420
19     366
20     308
21     259
22     214
23     178
24     147
25     124
26     100
27      79
28      66
29      48
30      41
31      28
32      21
33      20
34      18
35      17
36      13
37      10
38       9
39       7
40       4
41       4
42       2
43       2
44       1
45       1
46       1
Name: ID, dtype: int64
------------------
.            1653
,            1571
eta           712
\t            356
da            303
             ... 
ikerketa        1
onenaren        1
zauzkaten       1
keinuez         1
bidaliko        1
Name: FORM, Length: 8803, dtype: int64
------------------
.                1653
,                1571
izan             1431
*edun            1172
eta               737
                 ... 
laxatu              1
gurutzatu           1
e

In [19]:
np.unique(f0.UPOS.values)

array(['ADJ', 'ADP', 'ADV', 'AUX', 'CCONJ', 'DET', 'INTJ', 'NOUN', 'NUM',
       'PART', 'PRON', 'PROPN', 'PUNCT', 'SYM', 'VERB', 'X', '_'],
      dtype=object)

In [20]:
colnames

['ID',
 'FORM',
 'LEMMA',
 'UPOS',
 'XPOS',
 'FEATS',
 'HEAD',
 'DEPREL',
 'DEPS',
 'MISC']

In [21]:
def get_vals(f):
    cols = [ 'UPOS', 'XPOS', 'FEATS', 'HEAD', 'DEPREL', 'DEPS', 'MISC']
    dictvals = {}

    for n in cols:
        if not n in dictvals:
            try:
                dictvals[n] = set(np.unique(f[n].values))
            except:
                pass
        else:
            try:
                dictvals[n] = dictvals[n].union(set(np.unique(f[n].values) ) )
            except:
                pass
    return dictvals

In [22]:
get_vals(f0)

{'UPOS': {'ADJ',
  'ADP',
  'ADV',
  'AUX',
  'CCONJ',
  'DET',
  'INTJ',
  'NOUN',
  'NUM',
  'PART',
  'PRON',
  'PROPN',
  'PUNCT',
  'SYM',
  'VERB',
  'X',
  '_'},
 'XPOS': {'_'},
 'FEATS': {'1',
  '10',
  '11',
  '12',
  '13',
  '14',
  '15',
  '16',
  '17',
  '18',
  '19',
  '2',
  '20',
  '21',
  '22',
  '23',
  '24',
  '26',
  '28',
  '3',
  '30',
  '34',
  '39',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  'Animacy=Anim',
  'Animacy=Anim|Case=Abl',
  'Animacy=Anim|Case=Abl|Definite=Def|Number=Plur',
  'Animacy=Anim|Case=Abl|Definite=Def|Number=Sing',
  'Animacy=Anim|Case=Abs',
  'Animacy=Anim|Case=Abs|Definite=Def|Number=Plur',
  'Animacy=Anim|Case=Abs|Definite=Def|Number=Sing',
  'Animacy=Anim|Case=Abs|Definite=Ind',
  'Animacy=Anim|Case=All|Definite=Def|Number=Sing',
  'Animacy=Anim|Case=Ben|Definite=Def|Number=Plur',
  'Animacy=Anim|Case=Ben|Definite=Ind',
  'Animacy=Anim|Case=Com|Definite=Def|Number=Plur',
  'Animacy=Anim|Case=Com|Definite=Def|Number=Sing',
  'Animacy=Anim

In [23]:
allvals = []
errors = []
for cln in cleanfnames:
    try:
        f = pd.read_csv(cln, sep='\t', names=colnames)
        allvals.append(get_vals(f))
    except Exception as e:
        errors.append(cln)


/home/leo/venv3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [24]:
len(allvals)

316

In [25]:
len(errors)

23

In [26]:
errors

['/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Maltese-MUDT/mt_mudt-ud-dev.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Maltese-MUDT/mt_mudt-ud-test.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Maltese-MUDT/mt_mudt-ud-train.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Latvian-LVTB/lv_lvtb-ud-dev.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Latvian-LVTB/lv_lvtb-ud-train.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Italian-VIT/it_vit-ud-test.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Thai-PUD/th_pud-ud-test.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Slovak-SNK/sk_snk-ud-train.clean',
 '/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_Marathi-UFA

In [27]:
# for e in errors:
#     try:
#         f = pd.read_csv(e, sep='\t', names=colnames)
#     except Exception as e:
#         print(e)

In [28]:
# from deepmerge import Merger

# my_merger = Merger(
#     # pass in a list of tuple, with the
#     # strategies you are looking to apply
#     # to each type.
#     [
#         (set, ["union"]),
#         (list, ["append"]),
#         (dict, ["merge"])
#     ],
#     # next, choose the fallback strategies,
#     # applied to all other types:
#     ["override"],
#     # finally, choose the strategies in
#     # the case where the types conflict:
#     ["override"]
# )
# base = {"foo": ["bar"]}
# next = {"bar": "baz"}
# my_merger.merge(base, next)
# assert base == {"foo": ["bar"], "bar": "baz"}

In [29]:
# %%time
# allmerge = {}
# mergeerrors = []
# for v in allvals:
# #     print(type(v))
#     try:
#         my_merger.merge(allmerge,v)
#     except Exception as e:
#         mergeerrors.append(e)

In [30]:
# len(allmerge), len(mergeerrors)


In [31]:
# cols

In [32]:
%%time
allmerge = {}
for d in allvals:
#     print(type(d))
    for k,v in d.items():
        if k not in allmerge:
            allmerge[k] = v
        else:
            allmerge[k] = set(list(allmerge[k]) + list(v))
            

CPU times: user 23.2 s, sys: 3.83 ms, total: 23.2 s
Wall time: 23.2 s


In [33]:
allmerge.keys()

dict_keys(['UPOS', 'XPOS', 'FEATS', 'HEAD', 'DEPREL', 'DEPS', 'MISC'])

In [34]:
len(allmerge["UPOS"])

73

In [35]:
cols = [ 'UPOS', 'XPOS', 'FEATS', 'HEAD', 'DEPREL', 'DEPS', 'MISC']

In [36]:
all_upos = [d["UPOS"] for d in allvals if  "UPOS" in d]

In [37]:
upos = set.union(*all_upos)

There are issues parsing files this way, so I either have to fix these parser issues with pandas, or I can try another way of getting all the data from there with conllu lib

In [38]:
import conllu
import ntpath
# from https://stackoverflow.com/questions/8384737/extract-file-name-from-path-no-matter-what-the-os-path-format
def path_leaf(path):
    head, tail = ntpath.split(path)
    return tail or ntpath.basename(head)

In [39]:
#fname = "/home/leo/projects/Datasets/text/UniversalDependencies/ud-treebanks-v2.4/UD_French-PUD/fr_pud-ud-test.conllu"
def get_conllu(fname):
    f = open(fname,"r")
    cnl = conllu.parse(f.read())
    return path_leaf(fname), cnl

In [40]:
# fields according to conllu parser lib
#fields = ['id', 'form', 'lemma', 'upostag', 'xpostag', 'feats', 'head', 'deprel', 'deps', 'misc']
fields = ['upostag', 'xpostag', 'feats', 'deprel']

In [74]:
def filter_fields(conllu_dict, fields):
    """
    filter only the wanted fields from the conllu input dictionary
    """
    accum = {}
    for t in conllu_dict:  # each item is a line in a TokenList object
        for d in t:  # each item is a word/symbol
            for f in fields:  # only get the fields I need
                if f in accum:  #  and accum[f] is not None:
                    print(f, accum[f], d[f], accum[f], d[f])
                    accum[f] = accum[f].append(d[f])
                else:
                    lst = []
                    lst.append(d[f])
                    accum[f] = lst
    return accum

In [75]:
def get_conllu_tag_values(fname, fields):
    
    f = open(fname,"r")
    fid, parsed = get_conllu(fname)
    ret = (fid, filter_fields(parsed, fields))
    return ret


In [77]:
f = open(fname,"r")
fid, parsed = get_conllu(fname)

In [78]:
fid

'fr_pud-ud-test.conllu'

In [106]:
def get_tags(parsed):
    upostags = []
    xpostags = []
    feattags = []
    depreltags = []

    for t in parsed:
        for d in t:
            upostags.append(d['upostag'])
            xpostags.append(d['xpostag'])
            fts = []
            if d['feats'] is not None:
                fts = list(d['feats'].items())
                feattags.extend(fts)
            depreltags.append(d['deprel'])
            #feats are OrderedDicts
#             lfeattags = [(f.items) for f in feattags if f is not None]
    return set(upostags), set(xpostags), set(feattags), set(depreltags)

In [111]:
def get_conllu_tag_values(fname):
    f = open(fname,"r")
    fid, parsed = get_conllu(fname)
    ret = get_tags(parsed)
    return fid, ret

In [107]:
alltags = get_tags(parsed)

In [110]:
#alltags

In [113]:
%%time
all_tags = []
for cf in conllufiles:
    pid, tags = get_conllu_tag_values(cf)
    all_tags.append((pid,tags))

CPU times: user 6min 58s, sys: 7.19 s, total: 7min 5s
Wall time: 7min 5s


In [114]:
len(all_tags)

339

Now I'll do a union of all the sets, such as getting the basic values. FEATS is more complex and I'll deal with it later

In [116]:
upos = []
xpos = []
feats = []
deprel = []

for fid, tags in all_tags:
    upos.extend(tags[0])
    xpos.extend(tags[1])
    feats.extend(tags[2])
    deprel.extend(tags[3])

In [117]:
upos = set(upos)
xpos = set(xpos)
feats = set(feats)
deprel = set(deprel)

In [120]:
upos

{'ADJ',
 'ADP',
 'ADV',
 'AUX',
 'CCONJ',
 'DET',
 'INTJ',
 'NOUN',
 'NUM',
 'PART',
 'PRON',
 'PROPN',
 'PUNCT',
 'SCONJ',
 'SYM',
 'VERB',
 'X',
 '_'}

In [121]:
xpos

{'SN+SN+NR',
 'F1|grn1|casK|gen3|vgr2',
 'nnc+nbu+ncn',
 'Mlpmsi',
 '助詞-副助詞',
 'ClYS4----------',
 'Pron,Refl,Sg,Gen,PxPl2',
 'Td-po',
 'PZIP6----------',
 'PJFS6----------',
 'N,Sg,Ess,PxSg2',
 'M2|modM|tem4|grp1|casD|gen3',
 'VA+ETN+JKB',
 'V,Act,PcpVa,Sg,Ess',
 'ppas:pl:inst:m3:imperf:aff',
 'A,Qnt,Sg,Ade',
 'Va--1s----y',
 'PSFS1-P2-------',
 'NNip3',
 'Pfa--p',
 'AAns2y',
 'Ps1ms-s',
 'RhG-3PA--',
 'npp+ncpa+jca',
 'Pg-pnn',
 'v-ppp-mb-',
 'M3|modA|tem4|gen8|vgr2',
 'NNP+SN+NNG',
 'xsn+jco',
 'arfplyc',
 'NNG+NNG+NNB+JX',
 'VKdpb-',
 'vmyipt11san',
 'pvg+ep+etm',
 'PSIS7IS3-------',
 'cm',
 'Agsfpdy',
 'Ca--X----------',
 'C1|grn1|casF|gen3',
 'MM+NNG+NNB',
 'Vgma2s--n--ni-',
 'f+f+jcs',
 'Pda--p',
 'Ps3npgsf',
 'Pun,Quo,Fin',
 'paa+ep+etn+jca',
 'ADJ+NSUFF_MASC_DU_ACC_POSS',
 'ncpa+jxc+jca',
 'V,Neg,Act,Sg3,Ko',
 'Px-----ln',
 'SD----MD1-',
 'x--------',
 'Gtop1x',
 'Vag-----z',
 'm-p---fd-',
 'armsnyp',
 'ncn+ncpa+xsn',
 'Afpnsns',
 'C1|grn1|casJ|gen3',
 'Pphtds2',
 'NNP+XSN',
 

XPOS seems too much of a hassle to deal with in all the files for a complete multi-lang training. So I will discard it too for the moment.

In [123]:
len(deprel), deprel

(278,
 {'_',
  'acl',
  'acl:adv',
  'acl:appos',
  'acl:cleft',
  'acl:focus',
  'acl:inf',
  'acl:part',
  'acl:poss',
  'acl:relcl',
  'advcl',
  'advcl:appos',
  'advcl:arg',
  'advcl:cleft',
  'advcl:cond',
  'advcl:coverb',
  'advcl:periph',
  'advcl:relcl',
  'advcl:sp',
  'advcl:svc',
  'advcl:tcl',
  'advmod',
  'advmod:appos',
  'advmod:arg',
  'advmod:cc',
  'advmod:det',
  'advmod:df',
  'advmod:discourse',
  'advmod:emph',
  'advmod:locy',
  'advmod:mode',
  'advmod:neg',
  'advmod:obl',
  'advmod:periph',
  'advmod:que',
  'advmod:sentcon',
  'advmod:tfrom',
  'advmod:tlocy',
  'advmod:tmod',
  'advmod:to',
  'advmod:tto',
  'amod',
  'amod:advmod',
  'amod:att',
  'amod:attlvc',
  'amod:flat',
  'amod:mode',
  'amod:obl',
  'appos',
  'appos:conj',
  'appos:nmod',
  'aux',
  'aux:aglt',
  'aux:caus',
  'aux:clitic',
  'aux:cnd',
  'aux:imp',
  'aux:mood',
  'aux:neg',
  'aux:part',
  'aux:pass',
  'aux:poss',
  'aux:q',
  'case',
  'case:acc',
  'case:aspect',
  'case:ci

In [ ]:
# now do a parsing on the feats

In [124]:
len(feats)

610

In [125]:
feats

{('Abbr', 'Yes'),
 ('AdjType', 'Attr'),
 ('AdjType', 'Pred'),
 ('AdpType', 'Comadp'),
 ('AdpType', 'Comprep'),
 ('AdpType', 'Post'),
 ('AdpType', 'Prep'),
 ('AdpType', 'Preppron'),
 ('AdpType', 'Voc'),
 ('AdvType', 'Cau'),
 ('AdvType', 'Deg'),
 ('AdvType', 'Loc'),
 ('AdvType', 'Man'),
 ('AdvType', 'Mod'),
 ('AdvType', 'Sta'),
 ('AdvType', 'Tim'),
 ('Agglutination', 'Agl'),
 ('Agglutination', 'Nagl'),
 ('Analyt', 'Yes'),
 ('Animacy', 'Anim'),
 ('Animacy', 'Hum'),
 ('Animacy', 'Inan'),
 ('Animacy', 'Nhum'),
 ('Animacy[gram]', 'Anim'),
 ('Animacy[gram]', 'Inan'),
 ('Aspect', 'Dur'),
 ('Aspect', 'Dur,Iter'),
 ('Aspect', 'DurPerf'),
 ('Aspect', 'Hab'),
 ('Aspect', 'Imp'),
 ('Aspect', 'Imp,Iter'),
 ('Aspect', 'Imp,Perf'),
 ('Aspect', 'Inch'),
 ('Aspect', 'Iter'),
 ('Aspect', 'Iter,Perf'),
 ('Aspect', 'Perf'),
 ('Aspect', 'PerfRapid'),
 ('Aspect', 'Prog'),
 ('Aspect', 'ProgRapid'),
 ('Aspect', 'Prosp'),
 ('Aspect', 'Rapid'),
 ('Aspect', 'Res'),
 ('Case', 'Abe'),
 ('Case', 'Abl'),
 ('Case', 'A

Now I have to cut as many of these features as I can, so what I can do is:
- the ones that are a list, separate them
- the composite ones: separate them if feasible (separate by uppercase might work but not in all cases like Derivation)
- the sub-classes: if the sub-class has the same options as the class then just flatten it -> seems oversimplifying specially for languages that the same word can have several different meanings for the sub-classes ... )
    

In [126]:
n_feats = []
for k,v in feats:
    nv = v.split(',')
    nkv = zip([k]* len(nv), nv)
    n_feats.extend(nkv)

In [127]:
n_feats = set(n_feats)

In [128]:
len(n_feats)

521

In [131]:
n_feats1 = [f for f in n_feats if f[1]!='None']

In [132]:
len(n_feats1)

518

In [135]:
n_feats1 = sorted(n_feats1)

In [136]:
n_feats1

[('Abbr', 'Yes'),
 ('AdjType', 'Attr'),
 ('AdjType', 'Pred'),
 ('AdpType', 'Comadp'),
 ('AdpType', 'Comprep'),
 ('AdpType', 'Post'),
 ('AdpType', 'Prep'),
 ('AdpType', 'Preppron'),
 ('AdpType', 'Voc'),
 ('AdvType', 'Cau'),
 ('AdvType', 'Deg'),
 ('AdvType', 'Loc'),
 ('AdvType', 'Man'),
 ('AdvType', 'Mod'),
 ('AdvType', 'Sta'),
 ('AdvType', 'Tim'),
 ('Agglutination', 'Agl'),
 ('Agglutination', 'Nagl'),
 ('Analyt', 'Yes'),
 ('Animacy', 'Anim'),
 ('Animacy', 'Hum'),
 ('Animacy', 'Inan'),
 ('Animacy', 'Nhum'),
 ('Animacy[gram]', 'Anim'),
 ('Animacy[gram]', 'Inan'),
 ('Aspect', 'Dur'),
 ('Aspect', 'DurPerf'),
 ('Aspect', 'Hab'),
 ('Aspect', 'Imp'),
 ('Aspect', 'Inch'),
 ('Aspect', 'Iter'),
 ('Aspect', 'Perf'),
 ('Aspect', 'PerfRapid'),
 ('Aspect', 'Prog'),
 ('Aspect', 'ProgRapid'),
 ('Aspect', 'Prosp'),
 ('Aspect', 'Rapid'),
 ('Aspect', 'Res'),
 ('Case', 'Abe'),
 ('Case', 'Abl'),
 ('Case', 'Abs'),
 ('Case', 'Acc'),
 ('Case', 'Add'),
 ('Case', 'Ade'),
 ('Case', 'Advb'),
 ('Case', 'All'),
 ('C

In [ ]:
n_feats_simpl = []

In [137]:
518+18+278

814

It seems that there are 814 features that can be used, they are too many for me and my resources, so I should try to cut down a bit, but I don't know how for the moment. 
Maybe I just can try doing it with those many features (even if the loss computation will be slow, really slow) or just cut it to the upos features (they are just 18), then use the deprel field and later try with the feats that define more closely the elements.

Maybe just with the upos it already generates sufficient structure internally to be able to use it for translation, correction and generation.